runtime/src/global_request_scheduler_mts.c

```diff
@@ -0,0 +1,333 @@
+ */
+static struct sandbox_request *
+global_request_scheduler_mts_add(struct sandbox_request *sandbox_request)
+{
+	assert(sandbox_request);
+	assert(global_request_scheduler_mts_guaranteed && global_request_scheduler_mts_default);
+	if (unlikely(!is_this_listener_thread())) panic("%s is only callable by the listener thread\n", __func__);
+
+	struct module_global_request_queue *mgrq = sandbox_request->module->mgrq_requests;
+
+	LOCK_LOCK(&global_lock);
+
+	struct priority_queue *destination_queue = global_request_scheduler_mts_default;
+	if (sandbox_request->module->remaining_budget > 0) {
+		destination_queue = global_request_scheduler_mts_guaranteed;
+	}
+
+	uint64_t last_mrq_deadline = priority_queue_peek(mgrq->sandbox_requests);
+
+	int rc = priority_queue_enqueue_nolock(mgrq->sandbox_requests, sandbox_request);
+	if (rc == -ENOSPC) panic("Module's Request Queue is full\n");
+	// debuglog("Added a sandbox_request to the MGRQ");
+
+	/* Maintain the minheap structure by Removing and Adding the MGRQ from and to the global runqueue.
+	 * Do this only when the MGRQ's priority is updated.
+	 */
+	if (priority_queue_peek(mgrq->sandbox_requests) < last_mrq_deadline) {
+		priority_queue_delete_nolock(destination_queue, mgrq);
+
+		rc = priority_queue_enqueue_nolock(destination_queue, mgrq);
+		if (rc == -ENOSPC) panic("Global Runqueue is full!\n");
+		// debuglog("Added the MGRQ back to the Global runqueue - %s to Heapify", QUEUE_NAME);
+	}
+
+	LOCK_UNLOCK(&global_lock);
+
+	return sandbox_request;
+}
+
+/**
+ * @param pointer to the pointer that we want to set to the address of the removed sandbox request
+ * @returns 0 if successful, -ENOENT if empty
+ */
+int
+global_request_scheduler_mts_remove(struct sandbox_request **removed_sandbox_request)
+{
+	/* This function won't be used with the MTS scheduler. Keeping merely for the polymorhism. */
+	return -1;
+}
+
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_if_earlier(struct sandbox_request **removed_sandbox_request,
+                                                       uint64_t                 target_deadline)
+{
+	/* This function won't be used with the MTS scheduler. Keeping merely for the polymorhism. */
+	return -1;
+}
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @param mt_class the multi-tenancy class of the global request to compare the target deadline against
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_with_mt_class(struct sandbox_request **removed_sandbox_request,
```

Comment:
I can't really grok the logic in this function. It is very complex, and it is a testament to your abilities that you were able to get it to this state! If I were to proceed, I'd have to really reason through this, and potentially rewrite if there is any way to make it simpler. There might not be!!! Long story short, I think your ability to understand complexity is likely better than mine; sorry I'm not that useful here.